contracts/puppeteer-initia/src/contract.rs

```diff
@@ -0,0 +1,1161 @@
+    deps: DepsMut<NeutronQuery>,
+    info: MessageInfo,
+    validators: Vec<String>,
+) -> ContractResult<Response<NeutronMsg>> {
+    let puppeteer_base = Puppeteer::default();
+    let config = puppeteer_base.config.load(deps.storage)?;
+    cw_ownable::assert_owner(deps.storage, &info.sender)?;
+    cosmwasm_std::ensure!(
+        validators.len() < u16::MAX as usize,
+        StdError::generic_err("Too many validators provided")
+    );
+    deps.api.debug(&format!(
+        "WASMDEBUG: register_delegations_and_balance_query validators:{:?}",
+        validators
+    ));
+    let current_queries: Vec<u64> = puppeteer_base
+        .delegations_and_balances_query_id_chunk
+        .keys(deps.storage, None, None, Order::Ascending)
+        .collect::<StdResult<Vec<_>>>()?;
+    let messages = current_queries
+        .iter()
+        .map(|query_id| {
+            puppeteer_base
+                .delegations_and_balances_query_id_chunk
+                .remove(deps.storage, *query_id);
+            puppeteer_base.kv_queries.remove(deps.storage, *query_id);
+            NeutronMsg::remove_interchain_query(*query_id)
+        })
+        .collect::<Vec<_>>();
+
+    let mut submessages = vec![];
+    let ica = puppeteer_base.ica.get_address(deps.storage)?;
+
+    for (i, chunk) in validators
+        .chunks(config.delegations_queries_chunk_size as usize)
+        .enumerate()
+    {
+        submessages.push(SubMsg::reply_on_success(
+            new_delegations_and_balance_query_msg(
+                config.connection_id.clone(),
+                ica.clone(),
+                config.remote_denom.clone(),
+                chunk.to_vec(),
+                config.update_period,
+            )?,
+            ReplyMsg::KvDelegationsAndBalance { i: i as u16 }.to_reply_id(),
+        ));
+    }
+    Ok(Response::new()
+        .add_messages(messages)
+        .add_submessages(submessages))
+}
+
+fn execute_setup_protocol(
+    mut deps: DepsMut<NeutronQuery>,
+    env: Env,
+    info: MessageInfo,
+    delegate_grantee: String,
+    rewards_withdraw_address: String,
+) -> ContractResult<Response<NeutronMsg>> {
+    let puppeteer_base = Puppeteer::default();
+    let config: Config = puppeteer_base.config.load(deps.storage)?;
+    validate_sender(&config, &info.sender)?;
+    puppeteer_base.validate_tx_idle_state(deps.as_ref())?;
+    let ica = puppeteer_base.ica.get_address(deps.storage)?;
+    let mut any_msgs = vec![];
+    let grant_msg = MsgGrant {
+        grantee: delegate_grantee.clone(),
+        granter: ica.to_string(),
+        grant: Some(Grant {
+            authorization: Some(cosmos_sdk_proto::Any {
+                type_url: "/cosmos.authz.v1beta1.GenericAuthorization".to_string(),
+                value: GenericAuthorization {
+                    msg: "/initia.mstaking.v1.MsgDelegate".to_string(),
+                }
+                .encode_to_vec(),
+            }),
+            expiration: Some(prost_types::Timestamp {
```

Comment:
While we are here, can we provide None and make authz authorization infinite?

Reply:
no, we can't